sdk/cosmos/azure-cosmos-spark_3-5/src/main/scala/com/azure/cosmos/spark/ItemsWriterBuilder.scala

```diff
@@ -73,5 +85,84 @@ private class ItemsWriterBuilder
         sparkEnvironmentInfo)
 
     override def supportedCustomMetrics(): Array[CustomMetric] = supportedCosmosMetrics
+
+    override def requiredDistribution(): Distribution = {
+      if (writeConfig.bulkEnabled && writeConfig.bulkTransactional) {
+        // For transactional writes, partition by all partition key columns
+        val partitionKeyPaths = getPartitionKeyColumnNames()
+        if (partitionKeyPaths.nonEmpty) {
+          // Use public Expressions.column() factory - returns NamedReference
+          val clustering = partitionKeyPaths.map(path => Expressions.column(path): Expression).toArray
+          Distributions.clustered(clustering)
+        } else {
+          Distributions.unspecified()
+        }
+      } else {
+        Distributions.unspecified()
+      }
+    }
+
+    override def requiredOrdering(): Array[SortOrder] = {
+      if (writeConfig.bulkEnabled && writeConfig.bulkTransactional) {
+        // For transactional writes, order by all partition key columns (ascending)
+        val partitionKeyPaths = getPartitionKeyColumnNames()
+        if (partitionKeyPaths.nonEmpty) {
+          partitionKeyPaths.map { path =>
+            // Use public Expressions.sort() factory for creating SortOrder
+            Expressions.sort(
+              Expressions.column(path),
+              SortDirection.ASCENDING,
+              NullOrdering.NULLS_FIRST
+            )
+          }.toArray
+        } else {
+          Array.empty[SortOrder]
+        }
+      } else {
+        Array.empty[SortOrder]
+      }
+    }
+
+    private def getPartitionKeyColumnNames(): Seq[String] = {
+      try {
+        // Need to create a temporary container client to get partition key definition
+        val clientCacheItem = CosmosClientCache(
+          CosmosClientConfiguration(
+            userConfig.asCaseSensitiveMap().asScala.toMap,
+            ReadConsistencyStrategy.EVENTUAL,
+            sparkEnvironmentInfo
+          ),
+          Some(cosmosClientStateHandles.value.cosmosClientMetadataCaches),
+          "ItemsWriterBuilder-PKLookup"
+        )
+
+        val container = ThroughputControlHelper.getContainer(
+          userConfig.asCaseSensitiveMap().asScala.toMap,
+          containerConfig,
+          clientCacheItem,
+          None
+        )
+
+        val containerProperties = SparkBridgeInternal.getContainerPropertiesFromCollectionCache(container)
```

Comment:
You can simplify this by just using

  private val containerDefinition = SparkBridgeInternal
    .getContainerPropertiesFromCollectionCache(container)
  private val partitionKeyDefinition = containerDefinition.getPartitionKeyDefinition